### PR TITLE
Add Union.Bimap and some more InputUnion conversions

### PR DESCRIPTION
**Pull request by Frassle**
Resolves https://github.com/pulumi/pulumi/issues/8753

This adds conversion operators so that you can cast a `Union<T0,T1>` or
`Output<Union<T0,T1>>` to an `InputUnion<T0,T1>`. Without this
conversions it wasn't possible to create an `InputUnion` with a union
value, you could only give it a `T0` or `T1` (or an output of them).

This also adds `Union.Bimap` to easily convert both sides of the union,
it was possible to write this with match but Bimap is slightly nicer:
```
union.Match<Union<TResult0, TResult1>>(t0 => mapFunc0(t0), t1 => mapFunc1(t1)))

union.Bimap(mapFunc0, mapFunc1)
```

### CODE REVIEW

File: sdk/dotnet/Pulumi/Core/InputUnion.cs [hunk 1]
```diff
@@ -20,12 +21,18 @@ private InputUnion(Output<Union<T0, T1>> oneOf)
```
Comment (Frassle) [inline sdk/dotnet/Pulumi/Core/InputUnion.cs:21]: Note the two constructors above strict me as weird.
Why on earth do we allow `new InputUnion()` to make an empty union?! That looks like a great source for bugs.
Two why is the constructor from `Output<Union>` private? We've got a conversion operator for it now but maybe that should just be made public aswell/instead.

File: sdk/dotnet/Pulumi/PublicAPI.Shipped.txt [hunk 2]
```diff
@@ -322,8 +323,10 @@ static Pulumi.InputMap<V>.implicit operator Pulumi.InputMap<V>(Pulumi.Output<Sys
 static Pulumi.InputMap<V>.implicit operator Pulumi.InputMap<V>(Pulumi.Output<System.Collections.Immutable.ImmutableDictionary<string, V>> values) -> Pulumi.InputMap<V>
 static Pulumi.InputMap<V>.implicit operator Pulumi.InputMap<V>(System.Collections.Generic.Dictionary<string, V> values) -> Pulumi.InputMap<V>
 static Pulumi.InputMap<V>.implicit operator Pulumi.InputMap<V>(System.Collections.Immutable.ImmutableDictionary<string, V> values) -> Pulumi.InputMap<V>
+static Pulumi.InputUnion<T0, T1>.implicit operator Pulumi.InputUnion<T0, T1>(Pulumi.Output<Pulumi.Union<T0, T1>> value) -> Pulumi.InputUnion<T0, T1>
 static Pulumi.InputUnion<T0, T1>.implicit operator Pulumi.InputUnion<T0, T1>(Pulumi.Output<T0> value) -> Pulumi.InputUnion<T0, T1>
 static Pulumi.InputUnion<T0, T1>.implicit operator Pulumi.InputUnion<T0, T1>(Pulumi.Output<T1> value) -> Pulumi.InputUnion<T0, T1>
+static Pulumi.InputUnion<T0, T1>.implicit operator Pulumi.InputUnion<T0, T1>(Pulumi.Union<T0, T1> value) -> Pulumi.InputUnion<T0, T1>
```
Comment (mikhailshilkov) [inline sdk/dotnet/Pulumi/PublicAPI.Shipped.txt:329]: Would it make sense to add quick unit tests for these new APIs?

**Review by mikhailshilkov — APPROVED** [review]
Makes sense to me - added a few minor suggestions. cc @t0yv0 and @justinvp if they want to chime in.

File: CHANGELOG_PENDING.md [hunk 3]
```diff
@@ -1,5 +1,9 @@
 ### Improvements
 
+- [sdk/dotnet] Add `Union.Bimap` function for converting both sides of a union at once.
+  [#8733](https://github.com/pulumi/pulumi/pull/8733)
 
 ### Bug Fixes
 
+- [sdk/dotnet] Allow `Output<Union>` to be converted to `InputUnion`.
+  [#8733](https://github.com/pulumi/pulumi/pull/8733)
```
[the remark this answers is not in the document]
Comment (Frassle) [reply CHANGELOG_PENDING.md:9]: I didn't raise a github issue and I guess you could quibble about if it's a bug or not, but it's certainly surprising behavior and was blocking a user from some basic use of this type (see the linked slack thread).

File: sdk/dotnet/Pulumi/PublicAPI.Shipped.txt (hunk 2 above)
Comment (Frassle) [reply sdk/dotnet/Pulumi/PublicAPI.Shipped.txt:329]: Yes, I'll add some

File: CHANGELOG_PENDING.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikhailshilkov) [reply CHANGELOG_PENDING.md:17]: Opened https://github.com/pulumi/pulumi/issues/8753 for you

**Pull request merged**